### ISSUE TITLE
In MEMORY64 mode, function pointers are 64-bit, but should be 32-bit

### ISSUE THREAD

**Issue by juj** [emscripten-core/emscripten#18200]
Currently it looks like when compiling to -sMEMORY64, function pointers are 64-bit integers, and as result, e.g. {{{ makeDynCall() }}} macro does not currently work.

I cannot think of a technical reason why we would want to have 64-bit function pointers, since these will always be count-from-zero indexes to the wasm function table anyways. Would it be possible to make LLVM treat function pointers as 32-bit integers?

(soft reasons may be wanting to keep function pointers and integer pointers sizes consistent with each other, or to simplify e.g. some LLVM implementation)

**Comment by sbc100** [comment]
I think the reason for this is that having different size for function pointer and non-functions pointers is tricky.  I believe @aardappel tried to make it work but it wasn't possible.. I imagine from C/C++ language POV it could be hard/impossible to make that work.

On the emscripten side we handle 64-bit pointers at the boundary with JS by turning them into Numbers/int55.  I believe #18169 can/should be fixed by adding `__sig` to the relevant JS functions.

**Comment by aardappel** [comment]
Yup, this looked very complicated in LLVM if possible at all.. besides the fact that even if it would work, it would likely trip up all sorts of C/C++ code that make assumptions on all pointers being the same size.

I'd also say its fine to represent them as 53? bit integers in most places :)

**Comment by juj** [comment]
Thanks for the background. In the comment https://github.com/emscripten-core/emscripten/pull/18169/files#r1022043800 Sam mentions 

> The way we approach this issue with MEMORY64 is to tag all pointer-sized argument with p in the __sig.

> So this test function should have a corresponding `test_sig: "vpp"... then the value will be automatically converted to a Number/i53 as soon as it arrives on the JS side.

It is unfortunate to emit a `fnptr = Number(fnptr);` in the body of every JS function that takes in a function pointer. There are only two operations that are possible for a function pointer: call `getWasmTableEntry(fnptr)` on it, or pass it to another wasm function? 

So wouldn't it be better to have this `Number(fnptr)` casting inside `getWasmTableEntry()`? Then it would only need to exist in one location, and not in each JS library function that takes in a function pointer?
